win32ss/gdi/eng/pathdraw.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * PROJECT:     ReactOS kernel
+ * LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+ * PURPOSE:     Path drawing API.
+ * COPYRIGHT:   Copyright 2019 Katayama Hirofumi MZ.
+ */
+
+#include <win32k.h>
+#undef XFORMOBJ
+
+#define NDEBUG
+#include <debug.h>
+
+BOOL
+APIENTRY
+EngStrokePath(
+    IN SURFOBJ  *pso,
+    IN PATHOBJ  *ppo,
+    IN CLIPOBJ  *pco,
+    IN XFORMOBJ  *pxo,
+    IN BRUSHOBJ  *pbo,
+    IN POINTL  *pptlBrushOrg,
+    IN LINEATTRS  *plineattrs,
+    IN MIX  mix)
+{
+    // www.osr.com/ddk/graphics/gdifncs_4yaw.htm
+    UNIMPLEMENTED;
+    return FALSE;
+}
+
+/*
+ * @unimplemented
+ */
+BOOL
+APIENTRY
+EngFillPath(
+    IN SURFOBJ   *pso,
+    IN PATHOBJ   *ppo,
+    IN CLIPOBJ   *pco,
+    IN BRUSHOBJ  *pbo,
+    IN POINTL    *pptlBrushOrg,
+    IN MIX        mix,
+    IN FLONG      flOptions)
+{
+    // www.osr.com/ddk/graphics/gdifncs_9pyf.htm
+    UNIMPLEMENTED;
+    return FALSE;
+}
+
+BOOL
+APIENTRY
+EngStrokeAndFillPath(
+    IN SURFOBJ  *pso,
+    IN PATHOBJ  *ppo,
+    IN CLIPOBJ  *pco,
+    IN XFORMOBJ  *pxo,
+    IN BRUSHOBJ  *pboStroke,
+    IN LINEATTRS  *plineattrs,
+    IN BRUSHOBJ  *pboFill,
+    IN POINTL  *pptlBrushOrg,
+    IN MIX  mixFill,
+    IN FLONG  flOptions)
+{
+    // www.osr.com/ddk/graphics/gdifncs_2xwn.htm
+    UNIMPLEMENTED;
+    return FALSE;
+}
+
+__kernel_entry
+W32KAPI
+BOOL
+APIENTRY
+NtGdiEngStrokePath(
+    _In_ SURFOBJ *pso,
+    _In_ PATHOBJ *ppo,
+    _In_ CLIPOBJ *pco,
+    _In_ XFORMOBJ *pxo,
+    _In_ BRUSHOBJ *pbo,
+    _In_ POINTL *pptlBrushOrg,
+    _In_ LINEATTRS *plineattrs,
+    _In_ MIX mix)
+{
+    BOOL ret = FALSE;
+    POINT ptlBrushOrg;
+    LINEATTRS lineattrs;
+
+    _SEH2_TRY
+    {
+        ProbeForRead(pptlBrushOrg, sizeof(*pptlBrushOrg), 1);
+        ptlBrushOrg = *pptlBrushOrg;
+
+        ProbeForRead(plineattrs, sizeof(*plineattrs), 1);
+        lineattrs = *plineattrs;
+
+        ret = EngStrokePath(pso, ppo, pco, pxo, pbo, &ptlBrushOrg, &lineattrs, mix);
```

Comment:
This is not how it works. The usermode DDI structures cannot simply be used in kernel mode. Instead they need to be translated using UMPD objects.